docs/RFCS/split_sql_syntax.md

```diff
@@ -0,0 +1,138 @@
+- Feature Name: SQL Syntax for splitting tables
+- Status: draft/in-progress/completed/rejected/obsolete
+- Start Date: 2017-03-14
+- Authors: Radu Berinde
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #13665
+
+# Summary
+
+This RFC proposes new syntax for introducing table splits via SQL.
+
+# Motivation
+
+The main motivation is to allow setting up tests, benchmarks, and reproducible
+testbeds, especially for DistSQL. One set of some sample tests that we want to
+be able to set up directly from logic tests are in
+[TestDistSQLPlanner](https://github.com/cockroachdb/cockroach/blob/cc5ba57/pkg/sql/distsql_physical_planner_test.go#L75).
+These tests need to split tables in a specific way, and they need to reprogram
+the replicas and range leaders of each split.
+
+The secondary motivation is for backup/restore, which needs to introduce a
+specific set of key splits. Currently this is done by custom code which works as
+follows: the keys are sorted; a split is introduced on the middle key, then
+the splits to the left and respectively to the right are processed recursively
+(in parallel). We want the new syntax to implement this algorithm so that
+backup/restore can use it. Note that in this case, we already have the split keys
+(as opposed to primary key values). In addition to backup/restore, there are
+more usecases for pre-spliting ranges (for benchmarks, importing, etc).
+
+We currently have the `ALTER TABLE/INDEX SPLIT AT` statement which allows
+introducing a table or index split at specific column values. The main drawback
+is that it cannot be used *programmatically*: each split must be in its own
+statement; we cannot generate (e.g. via a `SELECT`) a set of values where to
+split. We want to write tests like the one linked above without hardcoding each
+split in the test file; and we want to be able to easily change the test table
+sizes. In addition, the `SPLIT AT` statments don't support control of
+replication.
+
+
+# Detailed design
+
+Three new pieces of SQL syntax are proposed.
+
+### 1. `key_prefix` function ###
+
+The `key_prefix` function is a function that takes a table name (optionally with
+an index) and a set of column values for a prefix of the primary key (or index
+columns) and returns the corresponding key (as `BYTES`). Example:
+
+```sql
+CREATE TABLE t (
+  k1 INT,
+  k2 INT,
+  v INT,
+  w INT,
+  PRIMARY KEY(k1, k2),
+  INDEX vw (v, w)
+)
+
+SELECT key_prefix('t', 1, 2)     -- returns /t/primary/1/2
```

Comment:
`key_prefix` just sounds too generic for what this is actually doing. Not sure what @petermattis had in mind with his comment, but I'd prefer renaming as follows:

`key_prefix` –> `distkv_key_prefix`
`pretty_key` –> `distkv_pretty_key`
`KVSPLIT` –> `DISTKV_SPLIT`
`KVRELOCATE` –> `DISTKV_RELOCATE`

This has the advantage of introducing a new namespace which is convincingly outside of any standard SQL dialect, while not being confusingly non-specific like `internal` or the like.